README
**Note - this is a defunct project and is purely here for historical reasons. The code is old, it's not great and I wouldn't use it in production (I'm not even sure if HellaNZB is still maintained). This is what I could import from the old SVN repository using svn2git.**
asd
HellaWorld v1.10-SVN
---------------------

*General Intall*

The installation is as simple as these 5 steps:

1) Copy config-sample.php to config.php
2) Open config.php in your favorite text editor
3) Edit the settings as you wish
4) Save and close your text editor
5) Upload the files to your webserver and view in your browser of choice

For more detailed configuration information, please read on.

*Authentication*

HellaWorld supports 4 authentication modes.

Open: Open to everyone without a password (Default)
Closed: Login required every time you visit
Hybrid: Open for specified IP ranges (defaults to the 3 standard private
        address ranges) login is required for all others
Exclusive: Closed for all addresses except those in the specified IP
           range

*Supported Languages*

Multiple languages are supported using GNU Gettext. Language codes are
in the form of lang_COUNTRY, so en_GB for English or sv_SE for Swedish.

Language codes with translations available are:
en_GB - English (Default)
en_US - American English
nl_NL - Dutch
sv_SE - Swedish

If you would like to translate HellaWorld into your native language,
please read the GNU Gettext documentation, the standard English messages
are located in locale/en_GB.UTF-8/messages.po

*Tracking Finished Items*

Tracking finished items will show a list of items completed by HellaNZB
but requires that both HellaNZB and HellaWorld be running on the same
machine, and does not work on Windows.

To track finished items you will have to edit your HellaNZB config to
include a Hellanzb.EXTERNAL_HANDLER_SCRIPT directive, like so:

Hellanzb.EXTERNAL_HANDLER_SCRIPT = '/path/to/hellaworld/completed.php'

For this to work, hellaworld/completed.xml must be writable by both the
user that HellaNZB is running as, and the user that HellaWorld is running
as. The latter is usually the same user that the web server is running as.

Displaying the finished items can be turned on or off in your config.php,
see the updates to config-sample.php to see what to add to existing
configurations.

Credits
---------------------
Many thanks for the suggestions, help and bug finding from:

Mat "Drarok" Gadd <http://www.drarok.com> for running safari and finding
all its javascript weaknesses for me, suggesting new features and helping
with general debugging.

Ludvig "toxik" Ericson <http://www.lericson.se> for various suggestions
for new features and modification of existing features.

"dannyp" in #hellanzb on freenode for suggesting the bookmarklet.

Many thanks to the FatBoyLan crew for providing me with a place to host my
site and Subversion repositories.
http://www.fatboylan.co.uk

$Id$


CHANGELOG
1.10

* Applied David Kubicek's patch for HTTP-uploading of NZBs, see
  http://trac.cheezyblog.net/ticket/21 for details. (drarok)

* Fixed a bug where descending sort wouldn't work (chris)

* Fixed a bug where queue items with an ETA over 24 but less than 48
  hours long weren't showing the days in the ETA (chris)

* Updated the bookmarklet to support Newzbin v2 and Newzxxx (chris)

1.9

* Fixed a bug where the total numbrsdfsdffer of iles and the number of NZB files
  processed wasn't showing. [ticket:5]

* Fixed a bug where the automatic refresh would reduce the log to a single
  line of text in IE. [ticket:4]

* Changed the display of timestamps to match HellaNZB, making them clearer

* Fixed a bug where the automatic refresh wouldn't update the final queue
  item when it began downloading [ticket:13]

* Dropped requirement for the Ctype library [ticket:12]

* Fixed a bug where finished items were still trying to be read despite
  the display of finished items being disabled. [ticket:15]

* Fixed a bug where only one processing item was showing after an auto
  refresh when HellaNZB was processing multiple items. [ticket:14]

* Added the ability to manually call the auto-refresh feature.

* Tidied up the appearance of HellaWorld when javascript is disabled.

* Altered the current, somewhat confusing, authentication system to a more
  intuitive configuration, also allowing for IP based authentication. This
  change has added options to the configuration file, please see the
  README for details

* Added further automatic refresh support for remaining data that wasn't
  already being refreshed by the 15 second automatic refresher [ticket:17]

1.8

* Fixed a Javascript bug where the 15 second refresh would start multiple
  refreshes after dragging and dropping. [ticket:2]

* Replaced some javascript with jquery shortcuts

* Made the Newzbin article ID box focused by default [ticket:3]

* Rewrote the sorting of the queue by name to use a more natural sorting
  algorithm, this means that, for example, "File 10" will come after
  "File 2" instead of before it.

* Fixed a bug where only one processing item would show when multiple
  items were being processed.

* Updated JQuery and its plugins to their latest versions, this will
  improve performance for large queues slightly.

* Added multi-call support to HellaController.

* Made sorting take advantage of the multi-call support, speeding up queue
  sorting significantly.

* Added gettext multi-language support. For information on translating
  HellaWorld to your language, please see the GNU Gettext documentation.

* Added partial translation into Dutch (Thanks Stevuu)

* Added partial translation into Swedish (Thanks Toxik)

1.7

* Added a workaround for a Javascript bug in Safari and Opera, which
  prevented the non javascript stylesheet from being disabled.

* Added a bookmarklet option, providing the ability to add NZBs to
  HellaNZB from a Newzbin article page without having to leave the page.

* Fixed a bug in the missing 'total_mb' field workaround. E_NOTICES should
  no longer be generated if HellaNZB fails to provide a size for a file.

* Fixed a Javascript bug where the queue wasn't updating on refresh.

* Added the ability to track finished items, see README for more info.

* Added support for SSL enabled webservers.

* Meta items have been broken up into tabs to reduce the amount of
  scrolling needed.

1.6 -- HellaWorld

* Name of project changed to HellaWorld due to similarity with HellaHella.

* Changed the way stylesheets are handled by non Javascript enabled clients
  to reduce the number of styles that need to be set at runtime by javascript
  enabled browsers

* Fixed a couple of minor CSS bugs in Internet Explorer

* Fixed a bug where NZB names weren't having special characters converted to
  html entities

* Fixed a bug in the error handling page. The error page no longer attempts to
  display HellaNZB status information when an error has occured, as under such
  conditions displaying the information is most likely impossible.

* Tidied up the javascript ever so slightly

* added HellaController::setRarPass() to HellaController

* Added the ability to specify a rar password for an NZB file from the queue.

* Fixed an uninitialized value bug in the authentication system caused by a
  typo

* Fixed a bug where adding a url instead of a newzbin id would have no effect

* Made sorting of the queue by name case insensitive

* Made the way HellaController::getStatus() handles the log entries more
  efficient

* Fixed a bug in HellaController::enqueue() where it was incorrectly
  parsing the input

* Tidied up HellaController:enqueueNewzbin() to clarify it

* Changed the input validation used by HellaController:enqueueURL() to
  make it less strict, and able to handle local URLs

* Changed the JSON template to use the PHP JSON extension.

* Fixed HellaController:listNZBs() so that it actually returns the list
  generated.

* Added handling to HellaController:getStatus() so that no E_NOTICEs are
  generated if HellaNZB doesn't return the size of a queue item

* Fixed a bug where the path wasn't being determined properly if
  HellaWorld was installed in a directory accessed via mod_alias

1.5
* Fixed a potential issue with the XML-RPC client, XML-RPC requests will now
  timeout before the maximum execution time is reached

* Added additional styles to the queue items so that they change colour on
  mouseover

* Fixed a bug where the option to refresh every 15 seconds was hidden if
  no items were in the queue

* Changed the re-order system from a select to a text box to reduce total
  HTML output, this will speed up large queues significantly. It should also
  help overcome a reoccuring bug in manual queue sorting where moving lots of
  queue items in one direction could lead to them being jumbled up.

* Moved from the bloated and slow script.aculo.us javascript framework to one

standard.php
<?php

/*
Copyright (c) 2007, Chris Stretton
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the names of The Cheezy Blog, Cheezyblog.net or
      HellaWorld nor the names of its contributors may be used to
      endorse or promote products derived from this software without
      specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
OF THE POSSIBILITY OF SUCH DAMAGE.

$Id$

*/

	function ipInRange($range, $address) {
		$range = str_replace(array(' ', "\r", "\n"), '', $range);
		list($a, $b, $c, $d) = explode('.', $address);
		$check = ($a << 24) + ($b << 16) + ($c << 8) + $d;
		foreach(explode(',', $range) as $ip) {
			list($base, $bits) = explode('/', $ip);
			list($a, $b, $c, $d) = explode('.', $base);

			$i = ($a << 24) + ($b << 16) + ($c << 8) + $d;
			$mask = $bits == 0 ? 0 : (~0 << (32 - $bits));
			$low = $i & $mask;
			$high = $i | (~$mask & 0xFFFFFFFF);

			if ($check >= $low && $check <= $high) {
				return true;
			}
		}
		return false;
	}

	try {

		if (file_exists('config.php')) {
			require_once 'config.php';
		} else {
			throw new Exception('config.php not found, please configure HellaWorld');
		}

		if (isset($config['auth']) && !empty($config['auth'])) {
			$auth = strtolower($config['auth']);
			if ($auth == 'open' || $auth == 'hybrid') {
				if (!isset($config['username']) || !isset($config['password']) || empty($config['username']) || empty($config['password'])) {
					throw new Exception('Authentication types open and hybrid require a password to be set');
				}
			}
		} else {
			$auth = 'open';
		}

		if (!function_exists('_')) {
			function _($string) {
				return $string;
			}
		} else {
			putenv('LC_ALL=' . $config['language'] . '.UTF-8');
			setlocale(LC_ALL, $config['language'] . '.UTF-8');
			bindtextdomain('messages', 'locale');
			textdomain('messages');
		}

		$hellaworldversion = "1.10-SVN";
		if (!isset($config['iprange'])) {
			$iprange = '192.168.0.0/16,10.0.0.0/8,172.168.0.0/12';
		} else {
			$iprange = $config['iprange'];
		}
		$protocol = (($_SERVER['SERVER_PORT'] == 443) ? 'https' : 'http');

		require_once 'classes/autoload.php';

		if ($auth == 'exclusive' && !ipInRange($iprange, $_SERVER['REMOTE_ADDR'])) {
			throw new Exception('IP Address is not in an allowed range');
		} elseif ($auth == 'closed' || ($auth == 'hybrid' && !ipInRange($iprange, $_SERVER['REMOTE_ADDR']))) {
			session_start();
			if (!array_key_exists('PHP_AUTH_USER', $_SERVER) || empty($_SERVER['PHP_AUTH_USER'])) {
				$_SESSION['login'] = 0;
				header('WWW-Authenticate: Basic realm="HellaWorld"');
				header('HTTP/1.0 401 Unauthorized');
				throw new Exception(_('Unauthorized access'));
			} else {
				$config['username'] = $_SERVER['PHP_AUTH_USER'];
				$config['password'] = $_SERVER['PHP_AUTH_PW'];
			}
			try {
				$c = new HellaController($config['host'], $config['port'], $config['username'], $config['password']);
			} catch (Exception $e) {
				if ($e->getCode() == 5) {
					$_SESSION['login'] ++;
					if ($_SESSION['login'] < 3) {
						header('WWW-Authenticate: Basic realm="HellaWorld"');
						header('HTTP/1.0 401 Unauthorized');
						throw new Exception(_('Unauthorized access'));
					} else {
						$_SESSION['login'] = 0;
						header('HTTP/1.0 403 Forbidden');
						throw new Exception(_('Access Forbidden'));
					}
				} else {
					throw new Exception($e->getMessage());
				}
			}
		} else {
			$c = new HellaController($config['host'], $config['port'], $config['username'], $config['password']);
		}
		$self = htmlentities($_SERVER['PHP_SELF']);

		if (array_key_exists('id', $_GET) && preg_match('/^\d+$/', $_GET['id'])) {
			$nzbid = $_GET['id'];
		} else {
			$nzbid = false;
		}
		if (array_key_exists('action', $_GET) && preg_match('/^\d+$/', $_GET['action'])) {
			switch ($_GET['action']) {
				case 1:
					$c->dequeue($nzbid);
					break;
				case 2:
					$c->up($nzbid);
					break;
				case 3:
					$c->down($nzbid);
					break;
				case 4:
					$c->first($nzbid);
					break;
				case 5:
					$c->last($nzbid);
					break;
				case 6:
					$c->force($nzbid);
					break;
				case 7:
					$c->pause();
					break;
				case 8:
					$c->resume();
					break;
				case 9:
					$c->cancel();
					break;
				case 10:
					$c->clear();
					break;
				case 11:
					$c->shutdown();
					throw new Exception('HellaNZB Has been shut down');
			}
			header('Location: ' . $protocol . '://' . $_SERVER['HTTP_HOST'] . $self);
		}

		$x = false;

		if (isset($config['showfinished']) && $config['showfinished'] && file_exists('completed.xml')) {
			$xmlstring = file_get_contents('completed.xml');
			if ($xmlstring) {
				$x = @new SimpleXMLElement($xmlstring);
			}
		}

		if (array_key_exists('refresher', $_GET) && preg_match('/^\d+$/', $_GET['refresher'])) {
			if ($_GET['refresher'] == 1) {
				include 'templates/json.php';
			}
			die();
		}

		if (array_key_exists('reorder', $_POST) && array_key_exists('sorttype', $_POST) && preg_match('/^\d+$/', $_POST['sorttype'])) {
			$type = $_POST['sorttype'];
			if ($type == 1 && array_key_exists('order', $_POST) && is_array($_POST['order'])) {
				arsort($_POST['order']);
				$c->multiCallStart();
				foreach($_POST['order'] as $queueindex => $neworder) {
					if (preg_match('/^\d+$/', $neworder) && $queueindex != $neworder) {
						$c->move($c->queue[$queueindex]['id'], $neworder);
					}
				}
				$c->multiCallCommit();
			} elseif (array_key_exists('sortdirection', $_POST) && preg_match('/^\d+$/', $_POST['sortdirection'])) {
				$sortdir = $_POST['sortdirection'];
				if ($type == 2) {
					$nzbName = array();
					foreach($c->queue as $queueitem) {
						$nzbName[] = $queueitem['nzbName'];
					}
					natcasesort($nzbName);
					if ($sortdir == 2) {
						$nzbName = array_reverse($nzbName, true);
					}
					$i = 1;
					$c->multiCallStart();
					while (list($id) = each($nzbName)) {
						$c->move($c->queue[$id]['id'], $i++);
					}
					$c->multiCallCommit();
				} elseif ($type == 3) {
					$size = array();
					foreach($c->queue as $queueitem) {
						$size[] = $queueitem['total_mb'];
					}
					natsort($size);
					if ($sortdir == 2) {
						$size = array_reverse($size, true);
					}
					$i = 1;
					$c->multiCallStart();
					while (list($id) = each($size)) {
						$c->move($c->queue[$id]['id'], $i++);
					}
					$c->multiCallCommit();
				}
			}
			header('Location: ' . $protocol . '://' . $_SERVER['HTTP_HOST'] . $self);
		}

		if (array_key_exists('queue', $_POST) && is_array($_POST['queue'])) {
			$tmp1 = array();
			$tmp2 = array();
			foreach($c->queue as $pos => $queue) {
				$tmp1[$pos] = $queue['id'];
				$tmp2[$pos] = substr($_POST['queue'][$pos], 6);
			}
			$diff = array_diff_assoc($tmp2, $tmp1);
			reset($diff);
			$fkey = key($diff);
			end($diff);
			$lkey = key($diff);
			if ($diff[$fkey] == $tmp1[$lkey]) {
				$c->move($diff[$fkey], $fkey + 1); // moving up
			} else {
				$c->move($diff[$lkey], $lkey + 1); // moving down
			}
			$c->getStatus();
			include 'templates/json.php';
			die();
		}

		if (array_key_exists('nzbdownload', $_GET)) {
			$nzbdownload = trim($_GET['nzbdownload']);
			if (preg_match('/^\d+$/', $nzbdownload)) {
				$c->enqueueNewzbin($nzbdownload);
			} else {
				$c->enqueueURL($nzbdownload);
			}
			header('Location: ' . $protocol . '://' . $_SERVER['HTTP_HOST'] . $self);
		}

		if (array_key_exists('nzbupload', $_FILES)) {
			$nzbfile = trim($_FILES['nzbupload']['tmp_name']);
			chmod($nzbfile, 0644);
			$nzbname = trim($_FILES['nzbupload']['name']);
			if (preg_match('/\.(nzb|zip)$/i', $nzbname)) {
				$newfile = dirname($nzbfile)."/".$nzbname;
				move_uploaded_file($nzbfile, $newfile);
				$c->enqueue($newfile);
				unlink($newfile);
			} else {
				throw new Exception("Invalid NZB upload!");
			}
			header('Location: ' . $protocol . '://' . $_SERVER['HTTP_HOST'] . $self);
		}

		if (array_key_exists('bookmarklet', $_GET)) {
			if (preg_match('/((https?):\/\/)?(([A-Z0-9][A-Z0-9_-]*)((\.[A-Z0-9][A-Z0-9_-]*)+)?)(:(\d+))?(\/([^ ]*)?| |$)/i', $_GET['bookmarklet']) < 1) {
				throw new Exception('Invalid bookmarklet URL');
			}
			$match = array();
			if (preg_match('/browse\/post\/(\d+)\/?$/', $_GET['bookmarklet'], $match) > 0) {
				$c->enqueueNewzbin($match[1]);
			}
			header('Location: ' . $_GET['bookmarklet']);
		}

		if (array_key_exists('maxrate', $_GET) && preg_match('/^\d+$/', $_GET['maxrate'])) {
			$c->setRate($_GET['maxrate']);
			header('Location: ' . $protocol . '://' . $_SERVER['HTTP_HOST'] . $self);
		}

		if (array_key_exists('info', $_GET) && preg_match('/^\d+$/', $_GET['info'])) {
			$index = $_GET['info'] - 1;
			include 'templates/rarpass.php';
			die();
		}

		if (array_key_exists('password', $_POST)) {
			if (PHP_VERSION < 6.0 && get_magic_quotes_gpc()) {
				$password = stripslashes($_POST['password']);
			} else {
				$password = $_POST['password'];
			}
			$c->setRarPass($nzbid, $password);
			$submitted = true;
			include 'templates/rarpass/php';
			die();
		}

		if (array_key_exists('clearnzbs', $_POST)) {
			if (file_exists('completed.xml')) {
				if (isset($x->item)) {
					unset($x->item);
					$fp = @fopen('completed.xml', 'w');
					if (!$fp) {
						throw new Exception(_('Unable to write to completed.xml, please check permissions'));
					}
					fwrite($fp, $x->asXML());
					fclose($fp);
				}
			}
			header('Location: ' . $protocol . '://' . $_SERVER['HTTP_HOST'] . $self);
		}

		if (array_key_exists('removefinished', $_GET) && preg_match('/^\d+$/', $_GET['removefinished']) && $x) {
			$id = --$_GET['removefinished'];
			if (isset($x->item[$id])) {
				unset($x->item[$id]);
				$fp = @fopen('completed.xml', 'w');
				if (!$fp) {
					throw new Exception(_('Unable to write to completed.xml, please check permissions'));
				}
				fwrite($fp, $x->asXML());
				fclose($fp);
				header('Location: ' . $protocol . '://' . $_SERVER['HTTP_HOST'] . $self);
			} else {
				throw new Exception(_('Invalid ID provided'));
			}
		}
		
		// Auto-NZB
		if (array_key_exists('addfeed', $_GET)) {
			$manager = new Feed_Manager(APPPATH.'feeds.xml');
			$manager->add($_GET['addfeed']);
		}

		// Have they ticked any "remove" boxes?
		if (array_key_exists('removefinished', $_POST) && is_array($remove = $_POST['removefinished']) && (bool) count($remove)) {
			// Reverse sort to avoid reindexing problems on the array.
			$remove = array_keys($remove);
			rsort($remove);

			foreach ($remove as $id)
				unset($x->item[$id - 1]);

			file_put_contents('completed.xml', $x->asXML());
		}

	} catch (Exception $e) {
		$errormessage = $e->getMessage();
		include 'templates/error.php';
		die();
	}

	include 'templates/default.php';
?>
